File: puzzle/yweiyst/solution/client-code/js/geom.js
class Rect {
	constructor(x, y, w, h) {
		this.x = x;
		this.y = y;
		this.w = w;
		this.h = h;
	}
	inBounds(x, y) {
		return x > this.x && x < this.x + this.w && y > this.y && y < this.y + this.h;
	}
};
Rect.ZERO = new Rect(0, 0, 0, 0);

class Vector3 {
	constructor(x, y, z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}
	dot(vec) {
		return this.x * vec.x + this.y * vec.y + this.z * vec.z;
	}
	static fromDict(d) {
		return new Vector3(
			d['x'], d['y'], d['z']
		);
	}
	toDict() {
		return {
			'x': this.x,
			'y': this.y,
			'z': this.z,
		}
	}
	equals(vec) {
		return this.x == vec.x && this.y == vec.y && this.z == vec.z;
	}
};

class CardinalDirection {
	static toAngle(d) {
		switch (d) {
		case CardinalDirection.N: return 0;
		case CardinalDirection.E: return Math.PI/2;
		case CardinalDirection.S: return Math.PI;
		case CardinalDirection.W: return 3*Math.PI/2;
		default:
			console.error('invalid CardinalDirection');
			return null;
		}
	}
	static isHorizontal(d) {
		return d == CardinalDirection.W || d == CardinalDirection.E;
	}
	static getOpposite(d) {
		switch (d) {
		case CardinalDirection.N: return CardinalDirection.S;
		case CardinalDirection.E: return CardinalDirection.W;
		case CardinalDirection.S: return CardinalDirection.N;
		case CardinalDirection.W: return CardinalDirection.E;
		default:
			console.error('invalid CardinalDirection');
			return null;
		}
	}
};

CardinalDirection.N = 0;
CardinalDirection.E = 1;
CardinalDirection.S = 2;
CardinalDirection.W = 3;
CardinalDirection.NUM = 4;

CardinalDirection.CLOCKWISE_FROM_N = [
	CardinalDirection.N,
	CardinalDirection.E,
	CardinalDirection.S,
	CardinalDirection.W,
];

class Matrix3x3 {
	constructor(x, y, z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}
	multVec(vec) {
		return new Vector3(
			this.x.dot(vec),
			this.y.dot(vec),
			this.z.dot(vec),
		);
	}
	getTranspose() {
		return new Matrix3x3(
			new Vector3(this.x.x, this.y.x, this.z.x),
			new Vector3(this.x.y, this.y.y, this.z.y),
			new Vector3(this.x.z, this.y.z, this.z.z)
		);
	}
	leftMult(mat) {
		const t = this.getTranspose();
		return new Matrix3x3(
			t.multVec(mat.x),
			t.multVec(mat.y),
			t.multVec(mat.z),
		);
	}
	// axis must be a unit vector, matrix must be ortho aa
	// rotation follows right-hand rule
	static makeRotate90DegOrthoAA(axis) {
		return new Matrix3x3(
			new Vector3(Math.abs(axis.x), -axis.z, axis.y),
			new Vector3(axis.z, Math.abs(axis.y), -axis.x),
			new Vector3(-axis.y, axis.x, Math.abs(axis.z))
		);
	}
	static fromDict(d) {
		return new Matrix3x3(
			Vector3.fromDict(d['x']),
			Vector3.fromDict(d['y']),
			Vector3.fromDict(d['z'])
		);
	}
	toDict() {
		return {
			'x': this.x.toDict(),
			'y': this.y.toDict(),
			'z': this.z.toDict(),
		}
	}
	equals(m) {
		return (
			this.x.equals(m.x) &&
			this.y.equals(m.y) &&
			this.z.equals(m.z)
		);
	}
};
Matrix3x3.IDENTITY = new Matrix3x3(
	new Vector3(1, 0, 0),
	new Vector3(0, 1, 0),
	new Vector3(0, 0, 1)
);

class CubeFace {
	static toVec(face) {
		switch (face) {
			case CubeFace.FRONT:
				return new Vector3(0, 0, 1);
			case CubeFace.BACK:
				return new Vector3(0, 0, -1);
			case CubeFace.LEFT:
				return new Vector3(-1, 0, 0);
			case CubeFace.RIGHT:
				return new Vector3(1, 0, 0);
			case CubeFace.TOP:
				return new Vector3(0, 1, 0);
			case CubeFace.BOTTOM:
				return new Vector3(0, -1, 0)
		}
		console.error('unknown CubeFace value ' + face.toString())
		return null;
	}
	static fromVec(vec) {
		if (vec.z == 1) return CubeFace.FRONT;
		if (vec.z == -1) return CubeFace.BACK;
		if (vec.x == -1) return CubeFace.LEFT;
		if (vec.x == 1) return CubeFace.RIGHT;
		if (vec.y == 1) return CubeFace.TOP;
		if (vec.y == -1) return CubeFace.BOTTOM;
		console.error('vector doesn\'t represent a CubeFace')
		return null;
	}
	static toRot(face) {
		return Matrix3x3.makeRotate90DegOrthoAA(CubeFace.toVec(face));
	}
};

CubeFace.FRONT = 0;
CubeFace.BACK = 1;
CubeFace.LEFT = 2;
CubeFace.RIGHT = 3;
CubeFace.TOP = 4;
CubeFace.BOTTOM = 5;
CubeFace.NUM = 6;

CubeFace.FRONT_VEC = CubeFace.toVec(CubeFace.FRONT);
CubeFace.BACK_VEC = CubeFace.toVec(CubeFace.BACK);
CubeFace.LEFT_VEC = CubeFace.toVec(CubeFace.LEFT);
CubeFace.RIGHT_VEC = CubeFace.toVec(CubeFace.RIGHT);
CubeFace.TOP_VEC = CubeFace.toVec(CubeFace.TOP);
CubeFace.BOTTOM_VEC = CubeFace.toVec(CubeFace.BOTTOM);

export { Rect, Vector3, Matrix3x3, CardinalDirection, CubeFace };

File: hunt/special_puzzles/ktane/geom.py
from enum import Enum

class Vector3():
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def dot(self, vec):
        return self.x * vec.x + self.y * vec.y + self.z * vec.z

    def from_dict(self, d):
        self.x = d['x']
        self.y = d['y']
        self.z = d['z']

    def to_dict(self):
        return {
            'x': self.x,
            'y': self.y,
            'z': self.z,
        }

    def __eq__(self, vec):
        return self.x == vec.x and self.y == vec.y and self.z == vec.z

class Matrix3x3():
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def mult_vec(self, vec):
        return Vector3(
            self.x.dot(vec),
            self.y.dot(vec),
            self.z.dot(vec),
        )

    def get_transpose(self):
        return Matrix3x3(
            Vector3(self.x.x, self.y.x, self.z.x),
            Vector3(self.x.y, self.y.y, self.z.y),
            Vector3(self.x.z, self.y.z, self.z.z)
        )

    def left_mult(self, mat):
        t = self.get_transpose()
        return Matrix3x3(
            t.mult_vec(mat.x),
            t.mult_vec(mat.y),
            t.mult_vec(mat.z),
        )

    @staticmethod
    def make_identity():
        return Matrix3x3(
            Vector3(1, 0, 0),
            Vector3(0, 1, 0),
            Vector3(0, 0, 1)
        )

    # axis must be a unit vector, matrix must be ortho aa
    # rotation follows right-hand rule
    @staticmethod
    def make_rotate_90deg_ortho_aa(axis):
        return Matrix3x3(
            Vector3(abs(axis.x), -axis.z, axis.y),
            Vector3(axis.z, abs(axis.y), -axis.x),
            Vector3(-axis.y, axis.x, abs(axis.z))
        )

    def from_dict(self, d):
        self.x.from_dict(d['x'])
        self.y.from_dict(d['y'])
        self.z.from_dict(d['z'])

    def to_dict(self):
        return {
            'x': self.x.to_dict(),
            'y': self.y.to_dict(),
            'z': self.z.to_dict(),
        }

    def __eq__(self, mat):
        return self.x == mat.x and self.y == mat.y and self.z == mat.z

# specialized for axis-aligned orthogonal matrices
class CardinalDirection(Enum):
    n = 0
    e = 1
    s = 2
    w = 3

    def move_point(self, x, y):
        return {
            CardinalDirection.n: (x, y-1),
            CardinalDirection.e: (x+1, y),
            CardinalDirection.s: (x, y+1),
            CardinalDirection.w: (x-1, y),
        }[self]

class CubeFace(Enum):
    front = 0
    back = 1
    left = 2
    right = 3
    top = 4
    bottom = 5

    def get_opposite(self):
        return {
            CubeFace.front: CubeFace.back,
            CubeFace.back: CubeFace.front,
            CubeFace.left: CubeFace.right,
            CubeFace.right: CubeFace.left,
            CubeFace.top: CubeFace.bottom,
            CubeFace.bottom: CubeFace.top,
        }[self]

    def get_vec(self):
        return {
            CubeFace.front: Vector3(0, 0, 1),
            CubeFace.back: Vector3(0, 0, -1),
            CubeFace.left: Vector3(-1, 0, 0),
            CubeFace.right: Vector3(1, 0, 0),
            CubeFace.top: Vector3(0, 1, 0),
            CubeFace.bottom: Vector3(0, -1, 0)
        }[self]

    @staticmethod
    def from_vec(vec):
        if vec.z == 1:
            return CubeFace.front
        elif vec.z == -1:
            return CubeFace.back
        elif vec.x == -1:
            return CubeFace.left
        elif vec.x == 1:
            return CubeFace.right
        elif vec.y == 1:
            return CubeFace.top
        elif vec.y == -1:
            return CubeFace.bottom
        else:
            raise Exception('vector doesn\'t represent a CubeFace')

    @staticmethod
    def to_rot(face):
        return Matrix3x3.make_rotate_90deg_ortho_aa(face.get_vec())

    @staticmethod
    def is_adjacent(face1, face2):
        return face1 != face2 and face1.get_opposite() != face2
